util/sparse_image_warp.py

```diff
@@ -0,0 +1,177 @@
+
+
+def _get_boundary_locations(image_height, image_width, num_points_per_edge):
+    """Compute evenly-spaced indices along edge of image."""
+    image_height = math_ops.to_float(image_height)
+    image_width = math_ops.to_float(image_width)
+    y_range = math_ops.linspace(0.0, image_height - 1, num_points_per_edge + 2)
+    x_range = math_ops.linspace(0.0, image_width - 1, num_points_per_edge + 2)
+    ys, xs = array_ops.meshgrid(y_range, x_range, indexing='ij')
+    is_boundary = math_ops.logical_or(
+                            math_ops.logical_or(math_ops.equal(xs, 0),                  # pylint: disable=bad-continuation
+                                                math_ops.equal(xs, image_width - 1)),
+                            math_ops.logical_or(math_ops.equal(ys, 0),                  # pylint: disable=bad-continuation
+                                                math_ops.equal(ys, image_height - 1)))
+    return array_ops.stack([array_ops.boolean_mask(ys, is_boundary),
+                            array_ops.boolean_mask(xs, is_boundary)], axis=-1)
+
+
+def _add_zero_flow_controls_at_boundary(control_point_locations,
+                                        control_point_flows, image_height,
+                                        image_width, boundary_points_per_edge):
+    """Add control points for zero-flow boundary conditions.
+    Augment the set of control points with extra points on the
+    boundary of the image that have zero flow.
+    Args:
+    control_point_locations: input control points
+    control_point_flows: their flows
+    image_height: image height
+    image_width: image width
+    boundary_points_per_edge: number of points to add in the middle of each
+                            edge (not including the corners).
+                            The total number of points added is
+                            4 + 4*(boundary_points_per_edge).
+    Returns:
+    merged_control_point_locations: augmented set of control point locations
+    merged_control_point_flows: augmented set of control point flows
+    """
+
+    batch_size = tensor_shape.dimension_value(control_point_locations.shape[0])
+
+    boundary_point_locations = _get_boundary_locations(image_height, image_width,
+                                                       boundary_points_per_edge)
+
+    boundary_point_flows = array_ops.zeros([array_ops.shape(boundary_point_locations)[0], 2])
+
+    boundary_point_locations = _expand_to_minibatch(boundary_point_locations,
+                                                    batch_size)
+
+    boundary_point_flows = _expand_to_minibatch(boundary_point_flows, batch_size)
+
+    merged_control_point_locations = array_ops.concat([control_point_locations, boundary_point_locations], 1)
+
+    merged_control_point_flows = array_ops.concat([control_point_flows, boundary_point_flows], 1)
+
+    return merged_control_point_locations, merged_control_point_flows
+
+
+def sparse_image_warp(image,
```

Comment:
Why inline this into the repo instead of using `tf.contrib.image.sparse_image_warp`? Was it modified?

Reply:
This is from where I got the code: https://github.com/tensorflow/tensorflow/pull/30904